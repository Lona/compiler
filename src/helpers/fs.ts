import * as fs from 'fs'
import * as path from 'path'

export type FSWrapper = {
  readFile(filePath: string): Promise<string>
  writeFile(filePath: string, data: string): Promise<void>
  copyDir(dirPath: string, output?: string): Promise<void>
}

export const createFSWrapper = (
  workspacePath: string,
  _outputPath?: unknown
): FSWrapper => {
  const outputPath =
    typeof _outputPath === 'string'
      ? _outputPath
      : path.join(process.cwd(), 'lona-generated')
  const fsWrapper = {
    readFile(filePath: string) {
      return fs.promises.readFile(
        path.resolve(workspacePath, filePath),
        'utf-8'
      )
    },
    writeFile(filePath: string, data: string) {
      const resolvedPath = path.resolve(outputPath, filePath)
      fs.mkdirSync(path.dirname(resolvedPath), { recursive: true })
      return fs.promises.writeFile(resolvedPath, data, 'utf-8')
    },
    async copyDir(dirPath: string, output: string = '.') {
      const resolvedPath = path.resolve(workspacePath, dirPath)
      const files = await fs.promises.readdir(resolvedPath)

      await Promise.all(
        files.map(async x => {
          if (
            (await fs.promises.stat(path.join(resolvedPath, x))).isDirectory()
          ) {
            return
          }

          return fsWrapper.writeFile(
            path.join(output, x),
            await fsWrapper.readFile(path.join(dirPath, x))
          )
        })
      )
    },
  }

  return fsWrapper
}
