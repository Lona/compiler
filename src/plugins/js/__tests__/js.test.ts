import * as path from 'path'
import * as fs from 'fs'
import Helpers, { Helpers as HelpersType } from '../../../helpers'
import * as formatter from '../'

const workspace = path.join(__dirname, '../../../../fixtures')

function readDir(x: string) {
  return fs
    .readdirSync(path.join(workspace, x))
    .map(y => path.join(workspace, x, y))
}

describe('JS', () => {
  let helpers: HelpersType

  beforeAll(async () => {
    helpers = await Helpers(workspace)
  })

  const files = readDir('declaration')
    .concat(readDir('expression'))
    .concat(readDir('literal'))
    .concat(readDir('statement'))

  files.forEach(filePath =>
    test(`Example ${filePath.split('fixtures/')[1]}`, async () => {
      const output = await formatter.convertFile(
        path.relative(workspace, filePath),
        helpers,
        {
          ...((helpers.config.format || {})[formatter.default.format] || {}),
        }
      )

      expect(output).toMatchSnapshot()
    })
  )
})
