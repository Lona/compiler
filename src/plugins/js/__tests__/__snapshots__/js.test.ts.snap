// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JS Example declaration/enumeration.md 1`] = `
"var OPTIONAL = { VALUE: \\"value\\", NONE: \\"none\\" }
"
`;

exports[`JS Example declaration/function.md 1`] = `
"var add = (a, b = 0) => {};
module.exports.add = add;
var add = (a, b = 0) => number.add(a, b) ;
module.exports.add = add;
var add = () => {};
module.exports.add = add;
"
`;

exports[`JS Example declaration/import.md 1`] = `
"
"
`;

exports[`JS Example declaration/namespace.md 1`] = `
"var boolean = { or: (a, b) => {}, and: (a, b) => {} };
module.exports.boolean = boolean;
"
`;

exports[`JS Example declaration/record.md 1`] = `
"
"
`;

exports[`JS Example declaration/variable.md 1`] = `
"var x = [ 42 ];
module.exports.x = x;
"
`;

exports[`JS Example expression/assignment.md 1`] = `
"var assignment = (a) => { a = 10 };
module.exports.assignment = assignment;
"
`;

exports[`JS Example expression/functionCall.md 1`] = `
"var functionCall = (a) => {
  { light: \\"pink\\", dark: \\"purple\\" }
  optional.none()
};
module.exports.functionCall = functionCall;
var colorIdentity = (a) => a ;
module.exports.colorIdentity = colorIdentity;
var pinkWithDetour = colorIdentity(\\"pink\\");
module.exports.pinkWithDetour = pinkWithDetour;
"
`;

exports[`JS Example expression/identifier.md 1`] = `
"var __lona_import_0 = require(\\"../declaration/variable\\");
var x = __lona_import_0.x;
var identifier = (a) => { a = x };
module.exports.identifier = identifier;
"
`;

exports[`JS Example expression/member.md 1`] = `
"var member = (a) => {
  a = colors.primary
  a.test = colors.primary
};
module.exports.member = member;
"
`;

exports[`JS Example literal/array.md 1`] = `
"var x = [ 42, 35 ];
module.exports.x = x;
"
`;

exports[`JS Example literal/boolean.md 1`] = `
"var trueBoolean = true;
module.exports.trueBoolean = trueBoolean;
var falseBoolean = false;
module.exports.falseBoolean = falseBoolean;
"
`;

exports[`JS Example literal/color.md 1`] = `
"var color = \\"pink\\";
module.exports.color = color;
var hexColor = \\"#123456\\";
module.exports.hexColor = hexColor;
"
`;

exports[`JS Example literal/number.md 1`] = `
"var positive = 1;
module.exports.positive = positive;
var float = 0.1;
module.exports.float = float;
var negative = -1;
module.exports.negative = negative;
"
`;

exports[`JS Example literal/string.md 1`] = `
"var string = \\"hello\\";
module.exports.string = string;
var stringWithQuote = \\"Hello \\\\\\"world\\\\\\"\\";
module.exports.stringWithQuote = stringWithQuote;
"
`;

exports[`JS Example statement/branch.md 1`] = `
"var branchStatement = () => {
  if (true) {
    return 0;
  }
  return 1;
};
module.exports.branchStatement = branchStatement;
"
`;

exports[`JS Example statement/loop.md 1`] = `
"var loopStatement = () => { while (false) {
    return 0;
  } };
module.exports.loopStatement = loopStatement;
"
`;

exports[`JS Example statement/return.md 1`] = `
"var returnStatement = () => 0 ;
module.exports.returnStatement = returnStatement;
"
`;
